File: find_shore.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <cmath>

struct Vertex { float x, y, z; };
std::vector<Vertex> verts;

void loadOBJ(const std::string& path) {
    std::ifstream file(path);
    std::string line;
    while (std::getline(file, line)) {
        if (line.substr(0, 2) == "v ") {
            std::istringstream s(line.substr(2));
            Vertex v; s >> v.x >> v.y >> v.z;
            verts.push_back(v);
        }
    }
}

float getTerrainHeight(float x, float z) {
    for (size_t i = 0; i < verts.size(); i += 3) {
        if (i + 2 >= verts.size()) break;
        Vertex v0 = verts[i], v1 = verts[i+1], v2 = verts[i+2];
        float minX = std::min(v0.x, std::min(v1.x, v2.x));
        float maxX = std::max(v0.x, std::max(v1.x, v2.x));
        float minZ = std::min(v0.z, std::min(v1.z, v2.z));
        float maxZ = std::max(v0.z, std::max(v1.z, v2.z));
        if (x >= minX && x <= maxX && z >= minZ && z <= maxZ) {
            float det = (v1.z - v2.z) * (v0.x - v2.x) + (v2.x - v1.x) * (v0.z - v2.z);
            float l1 = ((v1.z - v2.z) * (x - v2.x) + (v2.x - v1.x) * (z - v2.z)) / det;
            float l2 = ((v2.z - v0.z) * (x - v2.x) + (v0.x - v2.x) * (z - v2.z)) / det;
            float l3 = 1.0f - l1 - l2;
            if (l1 >= 0 && l2 >= 0 && l3 >= 0) return l1 * v0.y + l2 * v1.y + l3 * v2.y;
        }
    }
    return -1000.0f;
}

int main() {
    loadOBJ("Assets/models/strzeszynek/teren-jezioro.obj");
    for (float r = 50; r < 400; r += 20) {
        for (float a = 0; a < 360; a += 10) {
            float rad = a * M_PI / 180.0f;
            float x = r * cos(rad);
            float z = r * sin(rad);
            float h = getTerrainHeight(x, z);
            if (h >= 64.0f && h <= 66.0f) {
                std::cout << "Shore: " << x << ", " << h << ", " << z << "\n";
                return 0;
            }
        }
    }
}

File: find_shore.py
import struct
import math

with open('Assets/strzeszynek-teren-wys.raw', 'rb') as f:
    data = f.read()

width = 8192
height = 8192
max_dist = 400.0

def get_height(x, z):
    u = (x / max_dist) * 0.5 + 0.5
    v = (z / max_dist) * 0.5 + 0.5
    u = max(0.0, min(1.0, u))
    v = max(0.0, min(1.0, v))
    
    px = int(u * (width - 1))
    py = int((1.0 - v) * (height - 1))
    
    idx = (py * width + px) * 2
    if idx >= len(data) - 1: return -1000
    
    val = struct.unpack('<H', data[idx:idx+2])[0]
    normalized = val / 65535.0
    return normalized * 110.0 + 35.0

for radius in range(50, 400, 20):
    for angle in range(0, 360, 10):
        rad = math.radians(angle)
        x = radius * math.cos(rad)
        z = radius * math.sin(rad)
        h = get_height(x, z)
        if 64.0 <= h <= 65.0:
            print(f"Found shore at: x={x:.1f}, z={z:.1f}, h={h:.2f}")
            import sys
            sys.exit(0)
